lib/Service/RetentionService.php

```diff
@@ -0,0 +1,358 @@
+	protected int $userMaxLastLogin = 0;
+	protected int $guestDays = 0;
+	protected int $guestMaxLastLogin = 0;
+	protected array $excludedGroups = [];
+	protected array $reminders = [];
+	protected array $remindersPlain = [];
+	protected bool $keepUsersWithoutLogin = true;
+
+	public function __construct(
+		IConfig $config,
+		IUserManager $userManager,
+		IGroupManager $groupManager,
+		ITimeFactory $time,
+		IServerContainer $server,
+		IMailer $mailer,
+		IFactory $l10nFactory,
+		LoggerInterface $logger
+	) {
+		$this->config = $config;
+		$this->userManager = $userManager;
+		$this->groupManager = $groupManager;
+		$this->time = $time;
+		$this->server = $server;
+		$this->mailer = $mailer;
+		$this->l10nFactory = $l10nFactory;
+		$this->logger = $logger;
+	}
+
+	public function runCron(): void {
+		$now = new \DateTimeImmutable();
+		$this->userDays = (int) $this->config->getAppValue('user_retention', 'user_days', '0');
+		if ($this->userDays > 0) {
+			$userMaxLastLogin = $now->sub(new \DateInterval('P' . $this->userDays . 'D'));
+			$this->userMaxLastLogin = $userMaxLastLogin->getTimestamp();
+			$this->logger->debug('Account retention with last login before ' . $userMaxLastLogin->format(\DateTimeInterface::ATOM));
+		} else {
+			$this->logger->debug('Account retention is disabled');
+		}
+
+		$this->guestDays = (int) $this->config->getAppValue('user_retention', 'guest_days', '0');
+		if ($this->guestDays > 0) {
+			$guestMaxLastLogin = $now->sub(new \DateInterval('P' . $this->guestDays . 'D'));
+			$this->guestMaxLastLogin = $guestMaxLastLogin->getTimestamp();
+			$this->logger->debug('Guest account retention with last login before ' . $guestMaxLastLogin->format(\DateTimeInterface::ATOM));
+		} else {
+			$this->logger->debug('Guest account retention is disabled');
+		}
+
+		$reminderDaysString = $this->config->getAppValue('user_retention', 'reminder_days', '');
+		$reminderDayOptions = explode(',', $reminderDaysString);
+		foreach ($reminderDayOptions as $option) {
+			$option = (int) trim($option);
+			if ($option !== 0) {
+				$this->remindersPlain[] = $option;
+				$this->reminders[] = $now->sub(new \DateInterval('P' . $option . 'D'))->getTimestamp();
+			}
+		}
+
+		$this->keepUsersWithoutLogin = $this->config->getAppValue('user_retention', 'keep_users_without_login', 'yes') === 'yes';
+
+		try {
+			$excludedGroups = $this->config->getAppValue('user_retention', 'excluded_groups', '["admin"]');
+			$excludedGroups = json_decode($excludedGroups, true, 512, JSON_THROW_ON_ERROR);
+			$this->excludedGroups = \is_array($excludedGroups) ? $excludedGroups : [];
+		} catch (\JsonException $e) {
+		}
+
+		if ($this->keepUsersWithoutLogin) {
+			$this->userManager->callForSeenUsers(\Closure::fromCallable([$this, 'executeRetentionPolicy']));
+		} else {
+			$this->userManager->callForAllUsers(\Closure::fromCallable([$this, 'executeRetentionPolicy']));
+		}
+	}
+
+	public function executeRetentionPolicy(IUser $user): ?bool {
```

Comment:
```suggestion
	public function executeRetentionPolicy(IUser $user): bool {
```
Why return null?

Reply:
Documenting as specified on the consumer of the closure:
https://github.com/nextcloud/server/blob/8004aa77212eed6dd1142c713275a4e0c5d0c40e/lib/public/IUserManager.php#L202

Reply:
Yeah but you can be more specific on return type, as return type is covariant https://www.php.net/manual/en/language.oop5.variance.php :nerd_face:

Reply:
yeah, but would like to stay there to not have to revisit the poor documentation of the IUserManager 🙈